### DISCUSSION TITLE
Substitution of elFormEvents.addField/elFormEvents.removeField events with methods?

### DISCUSSION THREAD

**Discussion by purepear** [element-plus/element-plus#3158] [Ideas]
Big thanks for the incredible work on Element-plus!
Here is an idea that probably doesn't fit in the issues section :)

Maybe i'm missing something but i don't see a reason why we couldn't use simple methods instead of mitt events when we  add/remove form fields in a form. We already use `provide`/`inject` and the mitt events doesn't seem necessary. 🤔 Pls correct me if i'm wrong and this is used in some other ways or has a potential to be a breaking change. (Does the async nature of the events really help in this situation?)

In [form.vue](https://github.com/element-plus/element-plus/blob/dev/packages/components/form/src/form.vue#L118-L128) we have:
```
formMitt.on<FormItemCtx>(elFormEvents.addField, field => {
  if (field) {
    fields.push(field)
  }
})
formMitt.on<FormItemCtx>(elFormEvents.removeField, field => {
  if (field.prop) {
    fields.splice(fields.indexOf(field), 1)
  }
})
 ```
 That can become sth like:
 ```
 elForm.addField(field => {
  if (field) {
    fields.push(field)
  }
})
elForm.removeField(field => {
  if (field.prop) {
    fields.splice(fields.indexOf(field), 1)
  }
})
```
// We can  put these methods on the elForm `provide` and get rid of formMitt

---

And then in [form-item.vue](https://github.com/element-plus/element-plus/blob/dev/packages/components/form/src/form-item.vue#L345) we have:
```
onMounted(() => {
  if (props.prop) {
    elForm.formMitt?.emit(elFormEvents.addField, elFormItem)
  }
  ...  
})
onBeforeUnmount(() => {
  elForm.formMitt?.emit(elFormEvents.removeField, elFormItem)
})    
 ```
 That can become sth like:
 ```
 onMounted(() => {
  if (props.prop) {
    elForm.addField(elFormItem)
  }
  ...  
})
onBeforeUnmount(() => {
  elForm.removeField(elFormItem)
})    
 ```
 
 
 What do you think guys? Is it worth raising a PR? :)

**Comment by kooriookami** [comment]
@purepear Yes, remove `mitt` is a good feature. There is a PR before https://github.com/element-plus/element-plus/pull/431. If you interest in, you can refer that PR and improve it. PR is welcome. :D
@JeremyWuuuuu Do you have any suggestions?

**Comment by jw-foss** [comment]
@purepear It is definitely worth a PR for removing `mitt` actually as @kooriookami said, we had a community contributor that made a PR like this. But at that time, we were busy translating the codebase into `composition` APIs, so we left it there, but it would be fantastic to get rid of `mitt` for less dependent. If you are willing to do so, please feel free to. And if you had any trouble on doing this, please feel free to ping me or the team, we'd love to provide help! Thanks for making ElementPlus better. :heart:

**Comment by purepear** [comment]
Thanks guys, i'll look at the other PR and see if i can raise a new PR during the weekend :)

**Comment by purepear** [comment]
Here's what i came up with https://github.com/element-plus/element-plus/pull/3356 

All feedback appreciated. We can continue the discussion inside the PR comments 😄
